iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs

```diff
@@ -36,7 +37,7 @@ public ModuleAuthenticationWithHsm(ISignatureProvider signatureProvider, string
         protected override async Task<string> SafeCreateNewToken(string iotHub, int suggestedTimeToLive)
         {
             DateTime startTime = DateTime.UtcNow;
-            string audience = SasTokenBuilder.BuildAudience(iotHub, this.DeviceId, this.ModuleId);
+            string audience = SasTokenBuilder.BuildAudience(iotHub, WebUtility.UrlEncode(this.DeviceId), WebUtility.UrlEncode(this.ModuleId));
```

Comment:
Shouldn't this one be done by `BuildAudience` itself?

Reply:
Also, `BuildAudience` first constructs the `"{0}/devices/{1}/modules/{2}"` string and then `UrlEncode`s the whole thing. Should it be changed to encode the parts instead to be consistent with the other cases you've added?

Reply:
yeah, moved it to BuildAudience